config/config.go

```diff
@@ -22,105 +21,142 @@ const (
 
 // Field definition of a config value that can change.
 type Field struct {
-	Name    string
-	Kind    reflect.Kind
-	Version uint64
-	Sources map[Source]string
+	name    string
+	tp      string
+	version uint64
+	setter  reflect.Value
+	sources map[Source]string
 }
 
-// Config manages configuration and handles updates on the values.
-type Config struct {
-	Fields []*Field
-	sync.Mutex
-	cfg reflect.Value
-}
-
-// New creates a new monitor.
-func New(cfg interface{}) (*Config, error) {
-	if cfg == nil {
-		return nil, errors.New("configuration is nil")
+// NewField constructor.
+func NewField(fld *reflect.StructField, val *reflect.Value) (*Field, error) {
+	if !isTypeSupported(fld.Type) {
+		return nil, fmt.Errorf("field %s is not supported(only types from the sync package of harvester)", fld.Name)
 	}
-	tp := reflect.TypeOf(cfg)
-	if tp.Kind() != reflect.Ptr {
-		return nil, errors.New("configuration should be a pointer type")
+	f := &Field{
+		name:    fld.Name,
+		tp:      fld.Type.Name(),
+		version: 0,
+		setter:  val.FieldByName(fld.Name).Addr().MethodByName("Set"),
+		sources: make(map[Source]string),
 	}
-	ff, err := getFields(tp.Elem())
-	if err != nil {
-		return nil, err
+	value, ok := fld.Tag.Lookup(string(SourceSeed))
+	if ok {
+		f.sources[SourceSeed] = value
 	}
-	return &Config{cfg: reflect.ValueOf(cfg).Elem(), Fields: ff}, nil
+	value, ok = fld.Tag.Lookup(string(SourceEnv))
+	if ok {
+		f.sources[SourceEnv] = value
+	}
+	value, ok = fld.Tag.Lookup(string(SourceConsul))
+	if ok {
+		f.sources[SourceConsul] = value
+	}
+	return f, nil
+}
+
+// Name getter.
+func (f *Field) Name() string {
+	return f.name
+}
+
+// Type getter.
+func (f *Field) Type() string {
+	return f.tp
+}
+
+// Sources getter.
+func (f *Field) Sources() map[Source]string {
+	return f.sources
 }
 
-// Set the value of a property of the provided config.
-func (v *Config) Set(name, value string, kind reflect.Kind) error {
-	v.Lock()
-	defer v.Unlock()
-	f := v.cfg.FieldByName(name)
-	switch kind {
-	case reflect.Bool:
-		b, err := strconv.ParseBool(value)
+// Set the value of the field.
+func (f *Field) Set(value string, version uint64) error {
+	if version != 0 && version <= f.version {
+		return fmt.Errorf("version %d is older or same as field's %s version %d", version, f.name, f.version)
```

Comment:
You can also try this one if you think that it sounds better: `version %d is older or the same as the field's`.